Davonperson/chapter4 (#12)
mine-cetinkaya-rundel/r4ds-solutions: chapter-4.qmd

---
title: "Data transformation"
---

## Prerequisites

```{r}
library(nycflights13)
library(tidyverse)
#> ── Attaching packages ──────────────────────────────────── tidyverse 1.3.2 ──
#> ✔ ggplot2 3.4.0           ✔ purrr   1.0.1.9000 
#> ✔ tibble  3.1.8           ✔ dplyr   1.0.99.9000
#> ✔ tidyr   1.2.1.9001      ✔ stringr 1.5.0      
#> ✔ readr   2.1.3           ✔ forcats 0.5.2      
#> ── Conflicts ─────────────────────────────────────── tidyverse_conflicts() ──
#> ✖ dplyr::filter() masks stats::filter()
#> ✖ dplyr::lag()    masks stats::lag()
```

##Solutions

## 4.2.5 Exercises

> 1.  

Reviewer: Add some placeholder text like "Pipelines for each part are given below."

```{=html}
<!-- -->

> ```

Reviewer: Delete extraneous html block

a.  

```{r}

> filter(flights,arr_delay >= 2) |> 

Reviewer: start with `flighs` and pipe that into `filter()`

  arrange(desc(arr_delay))
```

> b.  Flew to houston (IAH or HOU)

Reviewer: Remove text

```{r}
flights |>

>   filter(dest %in% c("IAH","HOU"))

Reviewer: ```suggestion
  filter(dest %in% c("IAH", "HOU"))
```

```

c.  

```{r}
flights |>

>   filter(carrier %in% c("UA","DL","AA"))

Reviewer: ```suggestion
  filter(carrier %in% c("UA", "DL", "AA"))
```

```

d.  

```{r}
flights |>

>   filter(month %in% c(7,8,9))

Reviewer: Spaces after commas (throughout)

```

e.  

```{r}
flights |> 
  filter(dep_delay >= 0,arr_delay < -2)
```

f.  

```{r}
flights |> 
  filter(dep_delay <= -1,arr_delay > 30)
```

> 2.  

Reviewer: Add some placeholder text next to each question number.

```{r}
flights |> 

>   arrange(dep_delay) |> 

Reviewer: Question asks for longest delay so should use `desc(dep_delay)`

  arrange(time_hour)
```

3.  

```{r}
flights |> 

>   arrange(desc(arr_delay))

Reviewer: These are not the fastest flights, question asks to calculate speed.

```

> 4.Yes

Reviewer: ```suggestion
4.  Yes, there was a flight on every day of 2013.
```

```{r}
flights |> 
  arrange(desc(day)) |> 

>   arrange(desc(month))

Reviewer: This is not how we can determine that there was a flight on each day. Instead try

```r
flights |> 
  distinct(year, month, day)
```

and check to see if there are 365 unique dates.

```

5.  

```{r}
flights |> 
  arrange(distance) 
```

```{r}
flights |> 
  arrange(desc(distance))
```

> 6.  I don't think it matters the two functions would go over each other with no problems either way.

Reviewer: ```suggestion
6.  The order doesn't matter because we filter based on a condition, not based on row number.
```

## 4.3.5 Exercises

> 1.  No they don't match.

Reviewer: You should be checking if `sched_dep_time - dep_time` matches `dep_delay`.

>     ```{r}

Reviewer: Don't indent this chunk

    flights |> 
      mutate(timearrdep = arr_time - dep_time)
    ```

> 2.  They show when a plane left when it was supposed to leave and if there was a delay.

Reviewer: This doesn't address the question: Brainstorm as many ways as possible to select dep_time, dep_delay, arr_time, and arr_delay from flights.

3.  

```{r}
flights |> 
  select(dep_time, dep_delay, arr_time, arr_delay)
```

```{r}
select(flights, dep_time:arr_delay, -sched_dep_time,-sched_arr_time)
```

> 4.  Nothing happens, you just get the variable once.

Reviewer: This is the answer to Q3

```{r}
flights |> 
  select(dep_time, dep_time)
```

> 5.  You ask if `any_of()` these variables have a certain thing you are looking for.

Reviewer: Use `any_of()` in a pipeline to select any of these variables.

```{r}
variables <- c("year", "month", "day", "dep_delay", "arr_delay")
```

> 6.  No it doesn't surprise me.it looks like case doesn't matter.i can change time to dep.

Reviewer: This doesn't match the question, and Q7 is missing

```{r}
select(flights, contains("TIME"))
```

## 4.4.6 Exercises

> 1.  It looks like 9E has the worst delays.

Reviewer: Can you find out (by googling) which carrier that is?

```{r}
flights |> 
  group_by(carrier, dep_delay) |> 
  summarize(n())

flights |> 
  arrange(desc(dep_delay))
```

> 2.  

Reviewer: We're looking for most delayed by destination, so group by destination first and then arrange by dep_delay in decreasing order.

```{r}
flights |> 
  group_by(dep_delay,dest) |> 
  summarise(n())
```

3.  

```{r}

> #| eval: false

Reviewer: Why eval false?

ggplot(flights,aes(x = time_hour,y = dep_delay)) + 
  geom_smooth()
```

4.You get a error.

```{r}
#| error: true
flights |> 
  slice_min(-n)
```

> 5.Count,counts the number of things in each specific group,will show the largest group at the top.

Reviewer: ```suggestion
5.  `count()` counts the number of observations in each group, setting the `sort` argument to `TRUE` arranges the categories in descending order of number of observations.
```

6.  

```{r}
df <- tibble(
  x = 1:5,
  y = c("a", "b", "a", "a", "b"),
  z = c("K", "K", "L", "L", "K")
)
```

a.It would group by y.

```{r}
df |>
  group_by(y)
```

b.What arrange did here is arranged y in alphabetical order in this case.

```{r}
df |>
  arrange(y)
```

c.  The pipeline tells the code what to do next, it groups by y then finds the average of x.

```{r}
df |>
  group_by(y) |>
  summarize(mean_x = mean(x))
```

d.  The code groups by y and z then averages x,then it shows theres a average of 1 k for a y,an average of 3.5 L's for the second a in y and an average of 3.5 k's for b.

```{r}
df |>
  group_by(y, z) |>
  summarize(mean_x = mean(x))
```

e.  This code completely gets rid of the group by part the function that does this is .groups = "drop".

```{r}
df |>
  group_by(y, z) |>
  summarize(mean_x = mean(x), .groups = "drop")
```

> f.  These codes do the same thing the only difference is that the first code shows the information in a more efficient way due to summarize.

Reviewer: What is meant by efficient here? Consider using an alternative verb.
